### ISSUE TITLE
[unbound-method] False positive for Uncalled Function Checks

### ISSUE THREAD

**Issue by snebjorn** [typescript-eslint/typescript-eslint#1934]
**Repro**

```JSON
{
  "extends": [
      "airbnb-typescript/base",
      "plugin:@typescript-eslint/eslint-recommended",
      "plugin:@typescript-eslint/recommended",
      "plugin:@typescript-eslint/recommended-requiring-type-checking"
  ],
  "rules": {
  }
}
```

```TS
// https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#uncalled-function-checks
// this check is disallowed, but it's a reasonable function safety check aka uncalled-function-checks
// Avoid referencing unbound methods which may cause unintentional scoping of `this`. eslint@typescript-eslint/unbound-method
if (!!obj.doThing) {
  // safe function call, but linting error
  obj.doThing()
}
```

**Expected Result**

TypeScript advicates the `!!` syntax on functions
See: https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#uncalled-function-checks
> If you intended to test the function without calling it, you can correct the definition of it to include undefined/null, or use !! to write something like if (!!user.isAdministrator) to indicate that the coercion is intentional.

I'd expect the linting rules to match :)

**Actual Result**

eslint@typescript-eslint/unbound-method complains
> Avoid referencing unbound methods which may cause unintentional scoping of `this`. 

**Additional Info**

There is also an issue with the eslint rule `no-extra-boolean-cast` as it doesn't have type info. So it doesn't know that it's a function. Perhaps this exception can also dealt with?

> Redundant double negation. eslint(no-extra-boolean-cast)

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.29.0` |
| `@typescript-eslint/parser`        | `2.29.0` |
| `TypeScript`                       | `3.8.3` |
| `ESLint`                           | `6.8.0` |
| `node`                             | `13.12.0` |
| `npm`                              | `X.Y.Z` |
| `yarn`                              | `1.22.0` |

**Comment by snebjorn** [comment]
There's also an issue with this in (jest) tests.

```ts
expect(store._setState).not.toHaveBeenCalled();
```

Will result in
> Avoid referencing unbound methods which may cause unintentional scoping of `this`. eslint@typescript-eslint/unbound-method

**Comment by bradzacher** [comment]
For your second issue - that's unavoidable. According to the types, there is the possibility of that being an unbound method, so it must error.
I suggest you use eslint overrides to disable the rule on test files (#1776).

---

Otherwise - this is a bug indeed, it's very simple to fix - happily accept a PR.

**Comment by snebjorn** [comment]
The second issue was fixed in TSLint by allowing a whitelist https://github.com/palantir/tslint/pull/4472
Perhaps the same can be done here?

**Comment by snebjorn** [comment]
@bradzacher What about the `no-extra-boolean-cast` additional issue? Does typescript-eslint have the ability to conditionally turn off rules on sections?
If not I guess that means there needs to be a `typescript-eslint/no-extra-boolean-cast` version as eslint isn't able to handle this https://github.com/eslint/eslint/issues/13197

**Comment by bradzacher** [comment]
as in #1776, I don't think whitelists of names are a good idea in rules.
In rules that use type information, a whitelist creates a huge hole, because it lacks the type context.

EG, if you whitelist `expect` to stop your tests from erroring, then define an `expect` function in your local code, that function will never ever error due to the whitelist, which can introduce bugs into your code, and reduces trust in the linting tooling.

I purposely push back hard against whitelists unless there is literally no other suitable option.

----

`no-extra-boolean-cast` is a stylistic thing. Type information does not make sense here, because doing a `!!` isn't about specific types, it's about coercing any value to be a boolean.
If you're using that rule in your codebase, it means your codebase has chosen not to use `!!` to coerce a value to a boolean.

The solution is to use an explicit nullish check
```ts
if (someFunc != null) {
  someFunc();
}
```

or to just ditch the check and use optional chaining
```ts
someFunc?.();
```

**Comment by auhsoja** [comment]
I could probably fix this in the near future if no one else has claimed it.

**Comment by auhsoja** [comment]
I'm really sorry for potentially wasting more of your time, but is this known to be an issue in 2.30.0? I linted what I believe to be the exact same file with the exact same config and the following versions:

> package | version
> --- | ---
> `@typescript-eslint/eslint-plugin` | `2.30.0`
> `@typescript-eslint/parser`  | `2.30.0`
> `TypeScript` | `3.8.3`
> `ESLint` | `6.8.0`
> `yarn` | `1.22.0`

And the only error reported is the redundant double negation, which can't be fixed here if I'm not mistaken. Additionally, while I was looking through the codebase, it seemed as if an appropriate check was already in place in `unbound.ts`.

Thanks for your help and patience!

**Comment by bradzacher** [comment]
It errors for me against the current master:

```
class ContainsMethods {
  unbound?(): void;
}

let instance = new ContainsMethods();

if (!!instance.unbound) {
}
    

    assert.strictEqual(received, expected)

    Expected value to strictly be equal to:
      0
    Received:
      1
    
    Message:
      Should have no errors but had 1: [
      {
        ruleId: 'unbound-method',
        severity: 1,
        message: 'Avoid referencing unbound methods which may cause unintentional scoping of `this`.',
        line: 8,
        column: 7,
        nodeType: 'MemberExpression',
        messageId: 'unbound',
        endLine: 8,
        endColumn: 23
      }
    ]
```

**Comment by auhsoja** [comment]
I see. Sorry. I was being rather naive and just copied the original bug code without thinking about anything. In retrospect I'm more curious why it didn't complain about anything else. PR submitted.
